Add initial use cases for Enum generation

agustingomes opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | yes (at later stage)
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

This PR Enables the creation of Enum objects for the following use cases:

- Pure Enums
- Backed Enums either with string or integer values (as per RFC specification)

Fixes #100

agustingomes committed:
Add pure and backed Enum generation

Supports generating Enums that are pure or have backed values.

Signed-off-by: Agustin Gomes <me@agustingomes.com>

Ocramius commented on src/Generator/EnumGenerator.php:
```diff
@@ -0,0 +1,136 @@
+     */
+    private array $cases = [];
+
+    public function setName(string $name): self
```
On the new API, I wonder if we can leave out all setters as much as possible, and only use the constructor?

agustingomes replied:
I was thinking the same, using something like `EnumGenerator::fromArray()` to generate a new instance that is immutable sounds more solid.

weierophinney commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        $cases = [];
+        foreach ($backedCases as $case => $value) {
+            if ($type === 'string') {
```
Add a `&& is_string($value) || (is_object($value) && method_exists($value, '__toString'))` clause, and this error and the next should go away.

agustingomes replied:
I ended solving this error by tightening the docblock to the other allowed value, in this case, `int`

weierophinney commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,52 @@
+                $value = '\'' . $value . '\'';
+            }
+
+            $cases[] = $case . ' = ' . $value;
```
This makes it appear that the proper type for `$backedCases` is `array<string, string>`. Making that change may also correct the errors flagged i this method.

weierophinney commented on src/Generator/EnumGenerator/EnumGenerator.php:
```diff
@@ -0,0 +1,76 @@
+            return '';
+        }
+
+        return ': ' . $this->cases->getBackedType();
```
```suggestion
        return ': ' . $this->cases->getBackedType() ?: '';
```

weierophinney commented on src/Generator/EnumGenerator/Name.php:
```diff
@@ -0,0 +1,41 @@
+        if (false !== strpos($name, '\\')) {
+            $namespace = substr($name, 0, strrpos($name, '\\'));
+            $name      = substr($name, strrpos($name, '\\') + 1);
+        }
```
```suggestion
        $nsPosition = strrpos($name, '\\');
        if (false !== $nsPosition) {
            $namespace = substr($name, 0, $nsPosition);
            $name      = substr($name, $nsPosition + 1);
        }
```

This will correct the error, and prevent duplicate operations, making it perform faster.

weierophinney reviewed (changes requested):
The bulk of the static analysis errors are due to the engine needing to infer types, and not being able to do so. The way to correct that is to verify them before doing an operation. For values in arrays, check that the array key exists, and then that the value conforms to what is required, before performing an operation. In other cases, just make sure that the typehints provided are correct, particularly in annotations. If needed, uses webmozart/assert to do the validations.

agustingomes commented:
Thanks @weierophinney for the initial review. I'll keep your feedback in account as I work further in the PR

agustingomes commented on phpcs.xml:
```diff
@@ -9,6 +9,12 @@
     <exclude-pattern>*/TestAsset/*</exclude-pattern>
     <exclude-pattern>*/_files/*</exclude-pattern>
 
+    <rule ref="Generic.Files.LineLength.TooLong">
```
@Ocramius @weierophinney I added this ignore rule because the annotations for psalm I added were quite long.

weierophinney replied:
You can split annotations over multiple lines, especially those detailing array or object structures; I tend to prefer that approach as it has the side effect of making them more readable.

agustingomes replied:
That worked indeed. I pushed the change.

weierophinney commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,52 @@
+        $cases = [];
+        foreach ($backedCases as $case => $value) {
+            if ($type === 'string') {
+                $value = '\'' . $value . '\'';
```
I'd go with either:

```php
$value = "'" . $value . "'";
```

or:

```php
$value = sprintf("'%s'", $value);
```

to make it more clear that what you're doing is surrounding a string value with single quotes.

weierophinney commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+            return self::fromBackedCases($options['backedCases'] ?? ['type' => '', 'cases' => []]);
+        }
+
+        //TODO add more specific exception
```
```suggestion
        /** @todo add more specific exception */
```

weierophinney commented on test/Generator/EnumGeneratorTest.php:
```diff
@@ -0,0 +1,160 @@
+                'name'      => 'Suit',
+                'pureCases' => ['Hearts', 'Diamonds', 'Clubs', 'Spades'],
+            ],
+            <<<CODE
+enum Suit {
+    case Hearts;
+    case Diamonds;
+    case Clubs;
+    case Spades;
+}
+
+CODE,
```
```suggestion
            <<<CODE
                enum Suit {
                    case Hearts;
                    case Diamonds;
                    case Clubs;
                    case Spades;
                }

                CODE,
```

Since PHP 7.3, heredocs and nowdocs support indentation; any whitespace preceding the end delimiter will be stripped from the start of lines in the document. It's a nice feature to make inline heredoc/nowdocs more readable

agustingomes replied:
Aah that's cool, it's definitely much better the way you're suggesting.

weierophinney commented on test/Generator/EnumGeneratorTest.php:
```diff
@@ -0,0 +1,160 @@
+                'name'      => 'My\Namespaced\Suit',
+                'pureCases' => ['Hearts', 'Diamonds', 'Clubs', 'Spades'],
+            ],
+            <<<CODE
+namespace My\Namespaced;
+
+enum Suit {
+    case Hearts;
+    case Diamonds;
+    case Clubs;
+    case Spades;
+}
+
+CODE,
```
```suggestion
            <<<CODE
                namespace My\Namespaced;

                enum Suit {
                    case Hearts;
                    case Diamonds;
                    case Clubs;
                    case Spades;
                }

                CODE,
```

weierophinney commented on test/Generator/EnumGeneratorTest.php:
```diff
@@ -0,0 +1,160 @@
+                    'cases' => ['Hearts' => 'H', 'Diamonds' => 'D', 'Clubs' => 'C', 'Spades' => 'S'],
+                ],
+            ],
+            <<<CODE
+namespace My\Namespaced;
+
+enum Suit: string {
+    case Hearts = 'H';
+    case Diamonds = 'D';
+    case Clubs = 'C';
+    case Spades = 'S';
+}
+
+CODE,
```
```suggestion
            <<<CODE
                namespace My\Namespaced;

                enum Suit: string {
                    case Hearts = 'H';
                    case Diamonds = 'D';
                    case Clubs = 'C';
                    case Spades = 'S';
                }

                CODE,
```

weierophinney commented on test/Generator/EnumGeneratorTest.php:
```diff
@@ -0,0 +1,160 @@
+                    'cases' => ['Hearts' => 1, 'Diamonds' => 2, 'Clubs' => 3, 'Spades' => 4],
+                ],
+            ],
+            <<<CODE
+namespace My\Namespaced;
+
+enum Suit: int {
+    case Hearts = 1;
+    case Diamonds = 2;
+    case Clubs = 3;
+    case Spades = 4;
+}
+
+CODE,
```
```suggestion
            <<<CODE
                namespace My\Namespaced;

                enum Suit: int {
                    case Hearts = 1;
                    case Diamonds = 2;
                    case Clubs = 3;
                    case Spades = 4;
                }

                CODE,
```

weierophinney commented on test/Generator/EnumGeneratorTest.php:
```diff
@@ -0,0 +1,160 @@
+                    ],
+                ],
+            ],
+            <<<CODE
+enum Suit {
+    case Hearts;
+    case Diamonds;
+    case Clubs;
+    case Spades;
+
+    public function test() : string
+    {
+    }
+}
+
+CODE,
```
```suggestion
            <<<CODE
                enum Suit {
                    case Hearts;
                    case Diamonds;
                    case Clubs;
                    case Spades;

                    public function test() : string
                    {
                    }
                }

                CODE,
```

weierophinney reviewed (changes requested):
Few minor suggestions, mainly for readability. After making them, I think you can go ahead and remove the draft status, and @Ocramius can do final review.

agustingomes commented:
@weierophinney Thank you for the suggestions. I marked the PR as ready for review.

Ocramius commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,54 @@
+    private array $cases;
+
+    /**
+     * @param list<string> $cases
```
Are these `string` or `non-empty-string`?

agustingomes replied:
`non-empty-strings` is better here.

Ocramius commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,54 @@
+
+use function sprintf;
+
+final class BackedCases implements CasesInterface
```
Let's mark this `/** @psalm-immutable */`

Ocramius replied:
If I understand it correctly, `BackedCases` and `PureCases` are internal details, and should be marked `/** @internal */`?

agustingomes replied:
That is correct.

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * @psalm-param array{
+     *      name: string,
+     *      pureCases?: list<string>,
```
Are these `string` or `non-empty-string`?

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+     *      pureCases?: list<string>,
+     *      backedCases?: array{
+     *          type: string,
+     *          cases: array<string, int|string>
```
Are these `string` or `non-empty-string`?

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+     *      },
+     *      methods?: list<array{
+     *          name: string,
+     *          visibility?: string,
```
`visibility` as `string` seems a bit weak - can we improve on this, and document `'public'|'protected'|'private'`?

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+     *          cases: array<string, int|string>
+     *      },
+     *      methods?: list<array{
+     *          name: string,
```
Is this a `string` or `non-empty-string`?

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+     *          name: string,
+     *          visibility?: string,
+     *          static?: bool,
+     *          returntype?: string,
```
About `returntype`:

 * `returntype` or `returnType`?
 * could we expect a `TypeGenerator` instead of `string`?

agustingomes replied:
the property in the `MethodGenerator` is not camelcase, hence I left it like this.

I'll check the TypeGenerator suggestion.

Edit: the method setting the return type does not accept the `TypeGenerator` method. Is this something we have to change now, or can it be done in a separate PR?
![image](https://user-images.githubusercontent.com/15325886/142050747-a8941f77-4857-49aa-ae6f-3ed4bd981784.png)

agustingomes replied:
would `non-empty-string` work instead?

Ocramius replied:
> Edit: the method setting the return type does not accept the TypeGenerator method. Is this something we have to change now, or can it be done in a separate PR?

IMO we shouldn't expand more `type: string` API: it is a long-term liability.

Alternatively, completely drop this API, and remove the ability to use `::fromArray()` for this stuff.

agustingomes replied:
Yeah, I feel it's a good compromise. I'll drop the methods generation from this PR.

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+        $isPureCasesSpecified   = array_key_exists('pureCases', $options);
+        $isBackedCasesSpecified = array_key_exists('backedCases', $options);
+
+        if ($isPureCasesSpecified && $isBackedCasesSpecified === false) {
```
`$isBackedCasesSpecified === false` should imo be `! $isBackedCasesSpecified`, but it's just stylistic

agustingomes replied:
I tend to prefer adding the explicit comparison, as the `!` could be easily missed by the code readers at first glance.

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+        $isBackedCasesSpecified = array_key_exists('backedCases', $options);
+
+        if ($isPureCasesSpecified && $isBackedCasesSpecified === false) {
+            return self::fromPureCases($options['pureCases'] ?? []);
```
The `?? []` should not be needed here - we just verified that `array_key_exists()`

agustingomes replied:
Psalm complains if I omit that addition, and I'm not sure if it's because it's not understanding the check we're performing above.

Ocramius replied:
Potentially to be written as follows then (without variable):

```php
if (array_key_exists('backedCases', $options) ...) {
```

agustingomes replied:
that did the trick

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+            return self::fromPureCases($options['pureCases'] ?? []);
+        }
+
+        if ($isPureCasesSpecified === false && $isBackedCasesSpecified) {
```
`$isPureCasesSpecified === false` should be `! $isPureCasesSpecified` IMO

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+            return self::fromBackedCases($options['backedCases'] ?? ['type' => '', 'cases' => []]);
+        }
+
+        /** @todo add more specific exception */
```
To be solved before merge?

agustingomes replied:
Yes, I'll add a new exception. Do you have any suggestions what could we name it? `InvalidCasesSpecificationException` maybe?

Ocramius replied:
See comments around this: if we refine the types, the exception is not needed

agustingomes replied:
I tried to remove the exception thrown, but no luck. since it's possible for people in userland wrongly send options without neither property set, it seems we need to throw the exception when it happens. Do you see another way around this?

agustingomes replied:
never mind, using the `assert` above the return of the `backedCases` worked.

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+        }
+
+        /** @todo add more specific exception */
+        throw new Exception('Either `pureCases` or `backedCases` must be specified.');
```
We can avoid this scenario at type level.

The docblock would need to be exploded a bit:

```php
    /**
     * @psalm-param array{
     *      name: string,
     *      pureCases: list<string>,
     *      methods?: list<array{
     *          name: string,
     *          visibility?: string,
     *          static?: bool,
     *          returntype?: string,
     *      }>
     * }|array{
     *      name: string,
     *      backedCases: array{
     *          type: string,
     *          cases: array<string, int|string>
     *      },
     *      methods?: list<array{
     *          name: string,
     *          visibility?: string,
     *          static?: bool,
     *          returntype?: string,
     *      }>
     * } $options
     */
```

agustingomes replied:
But API consumers could still specify both properties in their implementations when using `EnumGenerator::fromConfig()`, wouldn't it be safer to throw the exception if that happens?

Even after implementing this dockblock, the analysis complains that the offsets of `pureCases` and `backedCases` could be undefined in their respective lines, even though we are checking beforehand.

Ocramius replied:
> could be undefined in their respective lines, even though we are checking beforehand.

Yes, that is because type inference can't understand if the left or right trees (in the suggested type above) are used.

That could be an `assert(array_key_exists(...))` after the first conditional:

```php
if (array_key_exists(...)) {
    return ...;
}

assert(array_key_exists(...));
return ...;
```

agustingomes replied:
indeed, moving them inline worked.

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    /**
+     * @param list<string> $pureCases
```
`non-empty-string` ?

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+
+    /**
+     * @psalm-param array{
+     *     type: string,
```
We should probably use `TypeGenerator` here

Ocramius commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,63 @@
+     */
+    public static function fromBackedCases(array $backedCases): BackedCases
+    {
+        return BackedCases::fromCasesWithType($backedCases['cases'] ?? [], $backedCases['type'] ?? '');
```
`$backedCases['type'] ?? ''` contradicts the declared type

Ocramius replied:
`$backedCases['cases'] ?? []` contradicts the declared type

Ocramius commented on src/Generator/EnumGenerator/Cases/PureCases.php:
```diff
@@ -0,0 +1,42 @@
+
+final class PureCases implements CasesInterface
+{
+    /** @var list<string> */
```
`non-empty-string`?

Ocramius commented on src/Generator/EnumGenerator/CasesInterface.php:
```diff
@@ -0,0 +1,13 @@
+
+namespace Laminas\Code\Generator\EnumGenerator;
+
+interface CasesInterface
```
This looks like a sealed type, where `CasesInterface = PureCases|BackedCases`

I suggest getting rid of the interface, and declaring the two types directly: this prevents implementing `CasesInterface` in userland, which would simplify our assumptions quite a bit.

agustingomes replied:
I agree, but that would require bumping the minimum required PHP version to 8.0, Should we do this in this PR?

agustingomes replied:
nevermind, doing it with DocBlocks is also good.

Ocramius commented on src/Generator/EnumGenerator/Cases/PureCases.php:
```diff
@@ -0,0 +1,42 @@
+
+    public function getBackedType(): ?string
+    {
+        return null;
```
If we remove the interface (as suggested below), and only use `PureCases|BackedCases` in our system, this nullability goes away

Ocramius commented on src/Generator/EnumGenerator/EnumGenerator.php:
```diff
@@ -0,0 +1,146 @@
+            $output .= 'namespace ' . $this->name->getNamespace() . ';' . self::LINE_FEED . self::LINE_FEED;
+        }
+
+        $output .= 'enum ' . $this->name->getName() . $this->retrieveType() . ' {'
```
Can return directly

Ocramius commented on src/Generator/EnumGenerator/EnumGenerator.php:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * @psalm-param array{
+     *      name: string,
+     *      pureCases?: list<string>,
```
The declaration of one of either `pureCases` or `backedCases` needing to exist should be ported here too, as shown in the `CaseFactory` suggestions

Ocramius commented on src/Generator/EnumGenerator/EnumGenerator.php:
```diff
@@ -0,0 +1,146 @@
+     * } $options
+     * @return list<MethodGenerator>
+     */
+    private static function fromMethods(array $options): array
```
This is not creating a `:self`, so the naming `::fromMethods()` is a bit confusing :thinking:

agustingomes replied:
would `::withProvidedMethodOptions()` work better?

Ocramius replied:
IMO no `with` in the naming: I would expect a `: self` return type, if it was called `with*` :) 

`private static function makeMethods(...): array` or such would fit better

Ocramius commented on src/Generator/EnumGenerator/Name.php:
```diff
@@ -0,0 +1,42 @@
+use function strrpos;
+use function substr;
+
+final class Name
```
Should probably be `/** @internal */` and have `/** @psalm-immutable */`

Ocramius commented on src/Generator/EnumGenerator/EnumGenerator.php:
```diff
@@ -0,0 +1,146 @@
+use function array_map;
+use function implode;
+
+final class EnumGenerator
```
This could potentially have a `/** @psalm-immutable */` attached

agustingomes replied:
Unfortunately we can't mark it as immutable because it's depending on the `MethodGenerator` and that class is not immutable :'(

Ocramius commented on src/Generator/EnumGenerator/EnumGenerator.php:
```diff
@@ -0,0 +1,146 @@
+
+        return $methodsOutput;
+    }
+}
```
One method missing here is `::fromReflection()`, operating on a given enum reflection instance - that's one of the biggest utilities in this generator system.

agustingomes replied:
I'll work on adding this method, but how can we do that while remaining compatible with 7.4?

From what I understand in the RFC, the reflection we need to use for that is `ReflectionEnum` but that is only included in 8.1.

michaelpetri replied:
Just an idea: https://3v4l.org/2SYBR

agustingomes replied:
I'll take a look at the suggestion

agustingomes replied:
I implemented the `::fromReflection()` method.

I updated the Psalm baseline for now, since I have no knowledge how to tackle the specific issues that arose from it, especially the `ParseError` ones, which makes sense since the analysis is done in 8.0.

Ocramius commented on src/Generator/EnumGenerator/Cases/PureCases.php:
```diff
@@ -0,0 +1,42 @@
+
+use function array_values;
+
+final class PureCases implements CasesInterface
```
`/** @psalm-immutable */`

Ocramius commented on test/Generator/EnumGeneratorTest.php:
```diff
@@ -0,0 +1,160 @@
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @group Laminas_Code_Generator
```
These groups can probably be dropped

michaelpetri commented on src/Generator/EnumGenerator/Name.php:
```diff
@@ -0,0 +1,46 @@
+ */
+final class Name
+{
+    private string $name;
+
+    private ?string $namespace;
```
Would both expect to be non-empty-string

agustingomes replied:
Existing implementations allow namespace to be omitted, so I assume we want to retain that characteristic for Enums.

michaelpetri replied:
Yes, sorry what i mean is `null|non-empty-string`.

agustingomes replied:
Since that property can still be `null`, I'll skip this suggestion.

michaelpetri commented on src/Generator/EnumGenerator/Name.php:
```diff
@@ -0,0 +1,46 @@
+        $this->namespace = $namespace;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    public function getNamespace(): ?string
+    {
+        return $this->namespace;
+    }
```
Since class is immutable, we can drop this.

agustingomes replied:
I'm not sure what do you mean here, you mean to remove the getters and make the properties public?

michaelpetri replied:
Yes exactly

weierophinney replied:
I'd wait until the minimum supported version on this is 8.1, and we can mark the properties readonly. Until then, it makes sense to have the properties private and use getters, as it prevents users from overwriting the values.

agustingomes replied:
I agree with @weierophinney 

In theory, any downstream project could abuse this object regardless that is annotated as `internal`. waiting for PHP 8.1 to become the minimum sounds good for me.

michaelpetri commented on src/Generator/EnumGenerator/Name.php:
```diff
@@ -0,0 +1,46 @@
+        return $this->namespace;
+    }
+
+    public static function fromName(string $name): self
```
Would rename to `fromFqcn`

agustingomes replied:
if it's deemed we must have a fully qualified class name, I'll adopt this suggestion.

michaelpetri replied:
I just think its a bit confusing since a property `name` exists, one could assume that the named ctor only sets the name. And since the fqcn of a class without a namespace is the class name itself it makes at least sense to me. But if you disagree just ignore it.

weierophinney replied:
I agree with `fromFqcn()` or `fromFullyQualifiedClassName()`. A global Enum with no namespace is fully qualified already, and hence would have no nuamespace associated; namespaced Enums will be split into a namespace + name. The name of the method should reflect this.

agustingomes replied:
I'll adopt the longer option since it's more explicit.

michaelpetri commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,57 @@
+        return $this->type;
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Class doesn't inherit from anything. But i wonder if it would make sense to introduce one in order to have a real type [here](https://github.com/laminas/laminas-code/pull/109/files#diff-a5540f944bc28e9bb26f0d15a7475d2f83b42347b9301f4b17bc9788ba2e104aR28) and [here](https://github.com/laminas/laminas-code/pull/109/files#diff-a5540f944bc28e9bb26f0d15a7475d2f83b42347b9301f4b17bc9788ba2e104aR33)

agustingomes replied:
Thanks for pointing it, will remove on the next push.

michaelpetri commented on src/Generator/EnumGenerator/Cases/CaseFactory.php:
```diff
@@ -0,0 +1,33 @@
+     * } $options
+     * @return BackedCases|PureCases
+     */
+    public static function fromOptions(array $options)
```
Related to the interface comment. But maybe instead of an interface we could introduce a abstract class and merge it with the factory:

```php
<?php

/** @internal */
abstract class Cases {
    /** @var list<non-empty-string> */
    private array $cases;

    /** @param list<non-empty-string> $cases */
    protected function __construct(array $cases)
    {
        $this->cases = $cases;
    }

    /** @psalm-return list<non-empty-string> */
    public function getCases(): array
    {
        return $this->cases;
    }

    /**
     * @psalm-param array{
     *      name: non-empty-string,
     *      pureCases: list<non-empty-string>,
     * }|array{
     *      name: non-empty-string,
     *      backedCases: array{
     *          type: 'int'|'string',
     *          cases: array<non-empty-string, int|non-empty-string>,
     *      },
     * } $options
     * @return BackedCases|PureCases
     */
    final public static function fromOptions(array $options): self
    {
        if (array_key_exists('pureCases', $options) && ! array_key_exists('backedCases', $options)) {
            return PureCases::fromCases($options['pureCases']);
        }

        assert(! array_key_exists('pureCases', $options) && array_key_exists('backedCases', $options));
        return BackedCases::fromCasesWithType($options['backedCases']['cases'], $options['backedCases']['type']);
    }
}

/**
 * @internal
 * @psalm-immutable
 */
final class PureCases extends Cases
{
    /** @psalm-param list<non-empty-string> $pureCases */
    public static function fromCases(array $pureCases): self
    {
        return new self($pureCases);
    }
}

/** 
 * @internal
 * @psalm-immutable  
 */
final class BackedCases extends cases
{
    private string $type;

    /** @param list<non-empty-string> $cases */
    private function __construct(string $type, array $cases)
    {
        $this->type  = $type;
        parent::__construct($cases);
    }

    public function getBackedType(): ?string
    {
        return $this->type;
    }

    /** @param array<string, string|int> $backedCases */
    public static function fromCasesWithType(array $backedCases, string $type): self
    {
        $cases = [];
        foreach ($backedCases as $case => $value) {
            if ($type === 'string') {
                $value = sprintf("'%s'", $value);
            }

            $cases[] = $case . ' = ' . $value;
        }

        return new self($type, $cases);
    }
}
```

agustingomes replied:
Personally I try to avoid abstract classes whenever possible, and for this particular case, I think putting the "factory" as part of the abstract class is not very extensible.

Next year, when PHP 7.4 is dropped (assuming that's the plan) we can use Union types instead.

michaelpetri commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    /**
+     * @param array<string, string|int> $backedCases
```
```suggestion
     * @param array<string, int>|array<string, string> $backedCases
```
Cases can't be mixed, Same [here](https://github.com/laminas/laminas-code/pull/109/files#diff-e301a27a5ae98f04f13cad35f3b4f5d41838e886deb3e0db9779fdc60967e84aR19) 

Also I really wonder if we need the second parameter when we know this?

agustingomes replied:
I'll add the suggestion, thanks 👍🏼 

As for the second param, I think having it is better than inferring the type from the array values.

weierophinney commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,57 @@
+        $this->cases = $cases;
+    }
+
+    public function getBackedType(): ?string
```
This will never be `null`, per the constructor and the declared property type.

weierophinney commented on src/Generator/EnumGenerator/Cases/BackedCases.php:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * @param array<string, int>|array<string, string> $backedCases
+     */
+    public static function fromCasesWithType(array $backedCases, string $type): self
```
We need validation on `$type`. Enums only allow for int or string types when using backed types.

agustingomes replied:
Will also add the respective exception here.

weierophinney reviewed (changes requested):
Getting close now! Thanks for all the work you're doing on this!

weierophinney reviewed (approved):
:+1:

Will wait for @Ocramius to do final review, however.

michaelpetri commented on test/Generator/EnumGeneratorTest.php:
```diff
@@ -0,0 +1,207 @@
+
+    public function testReflectionEnumFailsForUnsupportedPhpVersions(): void
+    {
+        if (PHP_VERSION_ID >= 80100) {
+            $this->markTestSkipped('ReflectionEnum is available from PHP 8.1 onwards.');
+        }
```
You could use `/** @requires PHP >= 8.1 */` on test methods

Ocramius replied:
5e6e3d013c32b3681533143b8dc7bc27fcd25d1d

michaelpetri commented on test/Generator/EnumGeneratorTest.php:
```diff
@@ -0,0 +1,207 @@
+
+    /**
+     * @dataProvider validEnumSpecifications
+     * @psalm-param array{
+     *      0: class-string,
+     *      1: non-empty-string
+     * }
```
```suggestion
     * @psalm-param class-string $enumClass
     * @psalm-param non-empty-string expected
```

Ocramius replied:
5e6e3d013c32b3681533143b8dc7bc27fcd25d1d

michaelpetri commented on psalm-baseline.xml:
```diff
@@ -1,5 +1,22 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <files psalm-version="4.10.0@916b098b008f6de4543892b1e0651c1c3b92cbfa">
+  <file src="polyfill/test/Environment.php">
+    <ParseError occurrences="1">
+      <code>Environment</code>
+    </ParseError>
+  </file>
+  <file src="polyfill/test/Flags.php">
+    <ParseError occurrences="2">
+      <code>Flags</code>
+      <code>case</code>
+    </ParseError>
+  </file>
+  <file src="polyfill/test/Orientation.php">
+    <ParseError occurrences="2">
+      <code>Orientation</code>
+      <code>case</code>
+    </ParseError>
+  </file>
```
For other 8.1 features i used eval to generate classes while runtime, see https://github.com/laminas/laminas-code/blob/4.5.x/test/Generator/PropertyGeneratorTest.php#L405

Ocramius replied:
Yeah, these will need to be moved to strings, for now: reason being that psalm chokes in a clean environment too.

Ocramius replied:
5e6e3d013c32b3681533143b8dc7bc27fcd25d1d

michaelpetri commented on psalm-baseline.xml:
```diff
@@ -258,6 +274,47 @@
+    <ArgumentTypeCoercion occurrences="1">
+      <code>$pureCases</code>
+    </ArgumentTypeCoercion>
+    <MixedArgument occurrences="2">
+      <code>$backingType-&gt;getName()</code>
+      <code>$enum-&gt;getCases()</code>
+    </MixedArgument>
+    <MixedArgumentTypeCoercion occurrences="1">
+      <code>$backedCases</code>
+    </MixedArgumentTypeCoercion>
+    <MixedArrayOffset occurrences="1">
+      <code>$backedCases[$singleCase-&gt;getName()]</code>
+    </MixedArrayOffset>
+    <MixedAssignment occurrences="3">
+      <code>$backedCases[$singleCase-&gt;getName()]</code>
+      <code>$backingType</code>
+      <code>$singleCase</code>
+    </MixedAssignment>
+    <MixedMethodCall occurrences="3">
+      <code>getBackingValue</code>
+      <code>getName</code>
+      <code>getName</code>
+    </MixedMethodCall>
+    <UndefinedClass occurrences="1">
+      <code>ReflectionEnumUnitCase</code>
+    </UndefinedClass>
+    <UndefinedMethod occurrences="3">
+      <code>getBackingType</code>
+      <code>getCases</code>
+      <code>getCases</code>
+    </UndefinedMethod>
+  </file>
+  <file src="src/Generator/EnumGenerator/EnumGenerator.php">
+    <MixedArgument occurrences="1">
+      <code>$enum-&gt;getName()</code>
+    </MixedArgument>
+    <UndefinedMethod occurrences="1">
+      <code>getName</code>
+    </UndefinedMethod>
+  </file>
```
Not sure if worth to do but you could extend the polyfill to get rid of this.

Ocramius replied:
IMO not worth doing here - these will disappear as we increase minimum PHP version, when viable.

Ocramius reviewed (approved):
Overall excellent:

 * the fact that implementations are `@internal` gives us space to improve later
 * everything else is simple and to the point

I'll proceed to merge, and will factor in the last comments by @michaelpetri  ( https://github.com/laminas/laminas-code/pull/109#pullrequestreview-824445421 )

Ocramius committed:
Adapted patch tests according to @michaelpetri's feedback

Ref: https://github.com/laminas/laminas-code/pull/109#pullrequestreview-824445421

In this change:

 * we remove the `polyfill/test` directory: we will use `<<<'PHP'` blocks directly
   in the test sources, in order to generate the classes that we need to verify
   enumerable type code generation
 * we replace `PHP_VERSION_ID`-based checks with PHPUnit's built-in `@requires PHP > 8.1`
   and similar
 * fixed some docblocks in tests
 * re-generated psalm baseline to reflect the fact that multiple issues were solved

Ocramius commented:
:ship: 

Thanks @agustingomes @michaelpetri @weierophinney! Gonna cut a release after dependency upgrades
